src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,302 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+
+    /// Class used to track call graph of a compilation
+    public class CallGraph
+    {
+        public struct CallGraphEdge
+        {
+            public ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType> ParamResolutions;
+        }
+
+        public struct CallGraphNode
+        {
+            public QsQualifiedName CallableName;
+            public QsSpecializationKind Kind;
+            public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
+        }
```

Comment:
🤔 Hmm, not sure about this one. I like these structs being tied to the CallGraph class in this way, and having outside code needing to use the qualified names as a reminder of what these are used for. In the future, these may even be private to the class. The structure of this whole thing will probably get changed at a later point.

Reply:
The [coding guidelines for .NET](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/nested-types) recommend avoiding nested types unless they are needed for access control (i.e. making a type visible only to a single class). Public nested types are discouraged (there's a gray squiggly in Visual Studio when you use them) because namespaces should be used for grouping instead of classes.

I think that requiring users of the class to write `CallGraph.CallGraphEdge` and `CallGraph.CallGraphNode` (or a static import) is unnecessary; even with `CallGraph.Edge` and `CallGraph.Node`, it's still unusual to need to use a dot in the type name (or a static import).